File: frontend/src/types/portfolio.ts
export interface PortfolioConstraints {
  max_single_asset_weight?: number;
  include_markets?: string[];
  include_asset_types?: string[];
}

export interface PortfolioGenerateRequest {
  risk_score: number;
  risk_tolerance: string;
  investment_horizon: string;
  strategy?: string;
  investment_amount?: number;
  currency?: string;
  constraints?: PortfolioConstraints;
}

export interface AssetSummary {
  symbol: string;
  name_ja: string | null;
  asset_type: string;
  market: string;
}

export interface AllocationResponse {
  asset: AssetSummary;
  weight: number;
  amount: number | null;
}

export interface PortfolioMetrics {
  expected_return: number | null;
  volatility: number | null;
  sharpe_ratio: number | null;
}

export interface RiskProfileSummary {
  risk_score: number;
  risk_tolerance: string;
}

export interface PortfolioResponse {
  name: string | null;
  strategy: string;
  risk_profile: RiskProfileSummary;
  metrics: PortfolioMetrics;
  allocations: AllocationResponse[];
  currency: string;
}

// Backtest
export interface BacktestAllocation {
  symbol: string;
  weight: number;
}

export interface BacktestRequest {
  allocations: BacktestAllocation[];
  period_years?: number;
  initial_investment?: number;
  rebalance_frequency?: string;
}

export interface BacktestPeriod {
  start: string;
  end: string;
  years: number;
}

export interface MaxDrawdownPeriod {
  start: string | null;
  end: string | null;
}

export interface BacktestMetrics {
  final_value: number;
  total_return: number;
  cagr: number;
  volatility: number;
  sharpe_ratio: number;
  max_drawdown: number;
  max_drawdown_period: MaxDrawdownPeriod | null;
  sortino_ratio: number | null;
  calmar_ratio: number | null;
}

export interface BenchmarkResult {
  total_return: number;
  cagr: number;
}

export interface TimeSeriesPoint {
  date: string;
  value: number;
  return_pct: number;
}

export interface AnnualReturn {
  year: number;
  return_pct: number;
}

export interface BacktestResponse {
  period: BacktestPeriod;
  initial_investment: number;
  metrics: BacktestMetrics;
  benchmark_comparison: Record<string, BenchmarkResult> | null;
  time_series: TimeSeriesPoint[];
  annual_returns: AnnualReturn[];
  disclaimer: string;
}

// Explain
export interface ExplainAllocationInput {
  symbol: string;
  name_ja?: string;
  weight: number;
}

export interface ExplainRequest {
  strategy: string;
  risk_tolerance: string;
  allocations: ExplainAllocationInput[];
  metrics?: Record<string, number>;
}

export interface ExplainResponse {
  explanation: string;
}

File: backend/src/app/schemas/portfolio.py
"""Portfolio-related Pydantic schemas."""

from pydantic import BaseModel, Field


# --- Request schemas ---

class PortfolioConstraints(BaseModel):
    max_single_asset_weight: float = 0.3
    include_markets: list[str] = ["jp", "us"]
    include_asset_types: list[str] = ["etf", "bond", "reit"]


class PortfolioGenerateRequest(BaseModel):
    risk_score: int = Field(..., ge=1, le=10)
    risk_tolerance: str = Field(..., pattern="^(conservative|moderate|aggressive)$")
    investment_horizon: str = Field(..., pattern="^(short|medium|long)$")
    strategy: str = "auto"
    investment_amount: int | None = None
    currency: str = "JPY"
    constraints: PortfolioConstraints | None = None


class AllocationInput(BaseModel):
    symbol: str
    weight: float = Field(..., ge=0, le=1)


class ExplainAllocationInput(BaseModel):
    symbol: str
    name_ja: str | None = None
    weight: float


class ExplainRequest(BaseModel):
    strategy: str
    risk_tolerance: str
    allocations: list[ExplainAllocationInput]
    metrics: dict | None = None


# --- Response schemas ---

class AssetSummary(BaseModel):
    symbol: str
    name_ja: str | None = None
    asset_type: str
    market: str


class AllocationResponse(BaseModel):
    asset: AssetSummary
    weight: float
    amount: float | None = None


class PortfolioMetrics(BaseModel):
    expected_return: float | None = None
    volatility: float | None = None
    sharpe_ratio: float | None = None


class RiskProfileSummary(BaseModel):
    risk_score: int
    risk_tolerance: str


class PortfolioResponse(BaseModel):
    name: str | None = None
    strategy: str
    risk_profile: RiskProfileSummary
    metrics: PortfolioMetrics
    allocations: list[AllocationResponse]
    currency: str


class ExplainResponse(BaseModel):
    explanation: str
